pkg/lib/authn/authenticator/password/generator.go

```diff
@@ -177,47 +173,70 @@ func (g *Generator) generate() (string, error) {
 	return shuffled, nil
 }
 
-func createSetFromList(list string) string {
-	var set strings.Builder
-	setMap := make(map[byte]bool)
+func (g *Generator) prepareCharList() string {
+	set := map[characterSet]struct{}{}
+
+	// Default to be alphanumeric.
+	set[characterSetAlphabet] = struct{}{}
+	set[characterSetDigit] = struct{}{}
 
-	for i := 0; i < len(list); i++ {
-		if _, ok := setMap[list[i]]; !ok {
-			set.WriteByte(list[i])
-			setMap[list[i]] = true
+	if g.Policy.LowercaseRequired {
+		set[characterSetLowercase] = struct{}{}
+	}
+	if g.Policy.UppercaseRequired {
+		set[characterSetUppercase] = struct{}{}
+	}
+	if g.Policy.AlphabetRequired && !g.Policy.LowercaseRequired && !g.Policy.UppercaseRequired {
+		set[characterSetAlphabet] = struct{}{}
+	}
+	if g.Policy.DigitRequired {
+		set[characterSetDigit] = struct{}{}
+	}
+	if g.Policy.SymbolRequired {
+		set[characterSetSymbol] = struct{}{}
+	}
+
+	var buf strings.Builder
+	for _, cs := range allCharacterSets {
+		if _, ok := set[cs]; ok {
+			cs.Append(&buf)
 		}
 	}
 
-	return set.String()
+	return buf.String()
 }
 
-func (g *Generator) prepareCharList() string {
-	charList := DefaultCharList
+func (g *Generator) getMinLength() int {
```

Comment:
Change this to a private function `func getMinLength(policy) int` and write tests for it.

Reply:
Updated together in [19a0230](https://github.com/authgear/authgear-server/pull/4535/commits/19a02307cb27dd2c508d83215277682e80115c74) 🙏